pkg/hubble/recorder/sink/dispatch.go

```diff
@@ -0,0 +1,231 @@
+	}
+
+	return &Dispatch{
+		startBootTime: bootTime,
+		startWallTime: wallTime,
+		sinkQueueSize: sinkQueueSize,
+		sinkByRuleID:  map[uint16]*sink{},
+	}, nil
+}
+
+// RegisterSink registers a new sink for the given rule ID. Any captures with a
+// matching rule ID will be forwarded to the pcap sink w. The provided header
+// is written to the pcap sink w upon initialization.
+func (d *Dispatch) RegisterSink(ctx context.Context, ruleID uint16, w pcap.RecordWriter, header pcap.Header) (*Handle, error) {
+	d.mutex.Lock()
+	defer d.mutex.Unlock()
+
+	if _, ok := d.sinkByRuleID[ruleID]; ok {
+		return nil, fmt.Errorf("sink for rule id %d already registered", ruleID)
+	}
+
+	s := startSink(ctx, w, header, d.sinkQueueSize)
+	d.sinkByRuleID[ruleID] = s
+	return &Handle{
+		C:    s.trigger,
+		sink: s,
+	}, nil
+}
+
+// UnregisterSink will stop and unregister the sink for the given ruleID.
+// It waits for any pending packets to be forwarded to the sink before closing
+// it and returns the final statistics or an error, if an error occurred.
+func (d *Dispatch) UnregisterSink(ctx context.Context, ruleID uint16) (stats Statistics, err error) {
+	d.mutex.Lock()
+	s, ok := d.sinkByRuleID[ruleID]
+	delete(d.sinkByRuleID, ruleID)
+	// unlock early to avoid holding the lock during s.close() which may block
+	d.mutex.Unlock()
+
+	if !ok {
+		return Statistics{}, fmt.Errorf("no sink found for rule id %d", ruleID)
+	}
+
+	if err = s.close(ctx); err != nil {
+		return Statistics{}, err
+	}
+
+	return s.copyStats(), nil
+}
+
+func (d *Dispatch) decodeRecordCaptureLocked(data []byte) (rec record, err error) {
+	dataLen := uint32(len(data))
+	if dataLen < monitor.RecorderCaptureLen {
+		return record{}, fmt.Errorf("not enough data to decode capture message: %d", dataLen)
+	}
+
+	// This needs to stay in sync with struct capture_msg from
+	// bpf/include/pcap.h.
+	// We could use binary.Read on monitor.RecorderCapture, but since it
+	// requires reflection, it is too slow to use on the critical path here.
+	const (
+		offsetRuleID         = 2
+		offsetBootTime       = 8
+		offsetCaptureLength  = 16
+		offsetOriginalLength = 20
+	)
+	n := byteorder.Native
+	ruleID := n.Uint16(data[offsetRuleID:])
+	bootTime := n.Uint64(data[offsetBootTime:])
+	capLen := n.Uint32(data[offsetCaptureLength:])
+	origLen := n.Uint32(data[offsetOriginalLength:])
+
+	// data may contain trailing garbage from the perf ring buffer
+	// https://lore.kernel.org/patchwork/patch/1244339/
+	packetEnd := monitor.RecorderCaptureLen + capLen
+	if dataLen < packetEnd {
+		return record{}, fmt.Errorf("capture record too short: want:%d < got:%d", dataLen, packetEnd)
+	}
+	packet := data[monitor.RecorderCaptureLen:packetEnd]
+
+	return record{
+		timestamp: d.bootTimeToWallTimeLocked(int64(bootTime)),
+		ruleID:    ruleID,
+		inclLen:   capLen,
+		origLen:   origLen,
+		data:      packet,
+	}, nil
+}
+
+func (d *Dispatch) bootTimeToWallTimeLocked(bootTime int64) time.Time {
```

Comment:
This function could be eliminated by moving the offset calculation to `getTimeNow()`.

Reply:
Fixed!